### PR TITLE
network: improve multipath route support

### LINKED ISSUE #17249: `Gateway=_dhcp` causes [Route] section ignored
**systemd version the issue has been seen with**
 > 246.6

**Used distribution**
 > Archlinux

**Linux kernel version used** (`uname -a`)
 > 5.8.13

**CPU architecture issue was seen on**
 > x64

**Expected behaviour you didn't see**
 > .network configured route is added

**Unexpected behaviour you saw**
 > error: `Route section without Gateway=, Destination=, Source=, or PreferredSource= field configured. Ignoring [Route] section from line xxx.`

**Steps to reproduce the problem**
 > `/etc/systemd/network/wired.network`
```
[Match]
Name=eth*

[Network]
DHCP=ipv4
LinkLocalAddressing=no
LLMNR=false

[DHCPv4]
UseDomains=true

[Route]
FastOpenNoCookie=true
Gateway=_dhcp
InitialAdvertisedReceiveWindow=88
InitialCongestionWindow=88
QuickAck=true
```

### PR DESCRIPTION
**Pull request by yuwata**
Also fixes #17249.

### CODE REVIEW

File: src/shared/local-addresses.c [hunk 1]
```diff
@@ -142,11 +142,46 @@ int local_addresses(sd_netlink *context, int ifindex, int af, struct local_addre
...
+        a = *list + *n_list;
+
+        *a = (struct local_address) {
+                .ifindex = ifindex,
+                .metric = metric,
+                .family = via->family,
+                .address = via->address,
+        };
+
+        (*n_list)++;
```
Comment (poettering) [inline src/shared/local-addresses.c]: maybe:

```c

`(*list)[(*n_list)++] = (struct local_address) {…
```

i.e. merge the las three lines in one? After all conceptually it's one operation: append an entry to the array...

File: src/network/networkd-address.h [hunk 2]
```diff
@@ -83,3 +83,9 @@ CONFIG_PARSER_PROTOTYPE(config_parse_lifetime);
 CONFIG_PARSER_PROTOTYPE(config_parse_address_flags);
 CONFIG_PARSER_PROTOTYPE(config_parse_address_scope);
 CONFIG_PARSER_PROTOTYPE(config_parse_duplicate_address_detection);
+
+#define ADDRESS_FMT_VAL(address)                   \
+        be32toh((address).s_addr) >> 24,           \
+        (be32toh((address).s_addr) >> 16) & 0xFFu, \
+        (be32toh((address).s_addr) >> 8) & 0xFFu,  \
+        be32toh((address).s_addr) & 0xFFu
```
Comment (poettering) [inline src/network/networkd-address.h:92]: what about the matching ADDRESS_FMT_STR macro? (we don't have this right now, but might make sense to have?

File: src/network/networkd-address.h [hunk 3]
```diff
@@ -83,3 +83,9 @@ CONFIG_PARSER_PROTOTYPE(config_parse_lifetime);
 CONFIG_PARSER_PROTOTYPE(config_parse_address_flags);
 CONFIG_PARSER_PROTOTYPE(config_parse_address_scope);
 CONFIG_PARSER_PROTOTYPE(config_parse_duplicate_address_detection);
+
+#define ADDRESS_FMT_VAL(address)                   \
```
Comment (poettering) [inline src/network/networkd-address.h]: maybe rename this IPV4_ADDRESS_FMT_VAL(), given that's what this is for only?

**Review by poettering — APPROVED** [review]
looks good to me, but i'd lie if i'd claim I grokked everything this does. two minor notes though

**Comment by yuwata** [comment]
@poettering Thank you for the review. 

> looks good to me, but i'd lie if i'd claim I grokked everything this does.

Haha.

**Comment by yuwata** [comment]
Hmm, RTA_VIA seems not supported by old kernels... Will adjust the test.

**Comment by kpfleming** [comment]
This PR added a warning when `Scope` is specified in a `Route` section and the target address is an IPv6 address, but did not update `man/systemd.network.xml` to reflect that `Scope` is no longer permitted in that section. Is that change coming in another PR, or should I open one?

**Pull request merged**